### PR TITLE
Infrastructure/alphabet cleanup

### PR DESCRIPTION
**Pull request by h-2**
* introduce semi-alphabet
* introduce proper concept documentation
* adapt some sub-modules

### CODE REVIEW

File: include/seqan3/alphabet/concept.hpp [hunk 1]
```diff
@@ -43,219 +43,81 @@
...
- * \tparam alphabet_type Must provide a `.to_char()` member function.
- * \param os The output stream you are printing to.
- * \param c The alphabet letter that you wish to convert to char.
- * \returns A reference to the output stream.
+/*!\interface seqan3::semi_alphabet_concept <>
+ * \brief The basis for seqan3::alphabet_concept, but requires only rank interface (not char).
+ * \ingroup alphabet
+ *
+ * \todo document
+ * This concept is mostly used internally and not relevant fo
```
Comment (smehringer) [inline include/seqan3/alphabet/concept.hpp]: fo... for what ?

File: include/seqan3/alphabet/concept.hpp [hunk 2]
```diff
@@ -43,219 +43,81 @@
...
-{
-    return c.to_char();
-}
-
-/*!\brief Ostream operator that delegates to `c.to_char()`.
- * \tparam alphabet_type Must provide a `.to_char()` member function.
- * \param os The output stream you are printing to.
- * \param c The alphabet letter that you wish to convert to char.
- * \returns A reference to the output stream.
+/*!\interface seqan3::semi_alphabet_concept <>
```
Comment (smehringer) [inline include/seqan3/alphabet/concept.hpp]: hm `semi_alphabet_concept` sounds weird to me.. ('semi' gives no clue, why the alphabet is unfinished or not an (real) alphabet.)
Maybe `rank_alphabet_concept`, `alphabet_base_concept` or `pre_alphabet_concept`..?

File: include/seqan3/alphabet/concept.hpp [hunk 3]
```diff
@@ -43,219 +43,81 @@
...
-{
-    return std::move(c.assign_char(in));
-}
-
-/*!\brief Free function wrapper that calls `.assign_rank(in)` on the first argument.
- * \tparam alphabet_type Must provide an `.assign_rank()` member function.
- * \param c The alphabet letter that you wish to assign to.
- * \param in The `rank` value you wish to assign.
- * \returns A reference to the alphabet letter you passed in.
+/*!\fn rank_type seqan3::to_rank(semi_alphabet_concept const a)
```
Comment (smehringer) [inline include/seqan3/alphabet/concept.hpp]: Why is there a function documentation with no function?  😖 Where is the function and why not document it directly?

Comment (smehringer) [inline include/seqan3/alphabet/concept.hpp]: I think we should avoid one character names. `a` has absolutely no meaning and whenever you want to find all occurrences in a (longer) code block, or you want to find and replace stuff, one charachter names are tedious.

File: include/seqan3/alphabet/concept.hpp [hunk 4]
```diff
@@ -43,219 +43,81 @@
...
- * \tparam alphabet_type Must provide an `.assign_rank()` member function.
- * \param c An alphabet letter temporary.
- * \param in The `rank` value you wish to assign.
- * \returns The assignment result as a temporary.
- * \details
- * Use this e.g. to newly create alphabet letters from rank:
- * ~~~{.cpp}
- * auto l = assign_rank(dna5{}, 1);  // l is of type dna5 and == dna5::C
- * ~~~
+/*!\fn semi_alphabet_concept && seqan3::assign_rank(semi_alphabet_concept && a, rank_type const i)
```
Comment (smehringer) [inline include/seqan3/alphabet/concept.hpp]: Again the one character names.. I would call them `letter` and `rank` (or `lttr`and `rnk`).

File: include/seqan3/alphabet/concept_detail.hpp [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    requires requires (alphabet_type a) { { a.to_char() } -> underlying_char_t<alphabet_type>; }
+{
+    return a.to_char();
+}
+
+/*!\brief Implementation of \link std::ostream & operator<<(std::ostream & os, alphabet_type const a) \endlink
+ * that delegates to `a.to_char()`.
+ * \tparam alphabet_type Must provide a `.to_char()` member function.
+ * \param os The output stream you are printing to.
+ * \param a The alphabet letter that you wish to convert to char.
```
Comment (smehringer) [inline include/seqan3/alphabet/concept_detail.hpp]: The alphabet letter you `wish to print.`

File: include/seqan3/alphabet/concept_detail.hpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+ * \link alphabet_concept && seqan3::assign_char(alphabet_concept && a, char_type const c)\endlink
+ * that delegates to a member function.
+ * \tparam alphabet_type Must provide an `.assign_char()` member function.
+ * \param a An alphabet letter temporary.
+ * \param c The `char` value you wish to assign.
+ * \returns The assignment result as a temporary.
+ * \details
+ * Use this e.g. to newly create alphabet letters from char:
+ * ~~~{.cpp}
+ * auto l = assign_char(dna5{}, 'G');  // l is of type dna5
```
Comment (smehringer) [inline include/seqan3/alphabet/concept_detail.hpp]: auto `dna5_letter` = ... 
I would rather give meaningful names.

File: include/seqan3/alphabet/concept_detail.hpp [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+template <typename alphabet_type>
+constexpr alphabet_type && assign_char(alphabet_type && a, underlying_char_t<alphabet_type> const c)
+    requires requires (alphabet_type a) { { a.assign_char(char{0}) } -> alphabet_type &; }
+{
+    return std::move(a.assign_char(c));
+}
+
+/*!\brief Implementation of
+ * \link semi_alphabet_concept && seqan3::assign_rank(semi_alphabet_concept && a, rank_type const i) \endlink
+ * that delegates to a member function.
```
Comment (smehringer) [inline include/seqan3/alphabet/concept_detail.hpp]: Actually, is it important for the user that the function "delegates to a member function" ? (Besides that this is obvious to see in the one line of code)
IMPOV we should inform the user about WHAT is happening in the \brief declaration and maybe HOW it is done in the details. What do you think?

File: include/seqan3/alphabet/concept_detail.hpp [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+ * \tparam alphabet_type Must provide an `.assign_rank()` member function.
+ * \param a An alphabet letter temporary.
+ * \param i The `rank` value you wish to assign.
+ * \returns The assignment result as a temporary.
+ * \details
+ * Use this e.g. to newly create alphabet letters from rank:
+ * ~~~{.cpp}
+ * auto l = assign_rank(dna5{}, 1);  // l is of type dna5 and == dna5::C
```
Comment (smehringer) [inline include/seqan3/alphabet/concept_detail.hpp]: Is there any documentation for each of the alphabets about which rank corresponds to which dna*_character?
If so, you should link it. If not it should be added :grimacing:.

File: include/seqan3/alphabet/nucleotide/concept.hpp [hunk 9]
```diff
@@ -68,11 +68,17 @@ constexpr bool is_nucleotide_v = is_nucleotide<type>::value;
 
 namespace seqan3
 {
-
-//!\brief A concept that indicates whether an alphabet represents nucleotides.
-//!\ingroup nucleotide
-//!\details Refines the seqan3::alphabet_concept.
+/*!\interface seqan3::nucleotide_concept <>
+ * \extends seqan3::alphabet_concept
+ * \brief A concept that indicates whether an alphabet represents nucleotides.
```
Comment (smehringer) [inline include/seqan3/alphabet/nucleotide/concept.hpp]: ... (e.g. `dna` or `dna5`). with links to the classes.
(I find it helpful, as often done in cppreference, to always include one example that satisfies this concept)
or is this maybe directly visible in doxygen? (..implemented by..)

File: include/seqan3/alphabet/concept.hpp (hunk 2 above)
Comment (h-2) [inline include/seqan3/alphabet/concept.hpp]: semi_* is a convention from the STL, see also http://en.cppreference.com/w/cpp/experimental/ranges/concepts/Semiregular

It also makes sense, because it has exactly half the functionality of an alphabet (all the rank stuff, but not the char stuff).

File: include/seqan3/alphabet/concept.hpp (hunk 1 above)
Comment (h-2) [inline include/seqan3/alphabet/concept.hpp]: oops, I forgot this, I will update it.

File: include/seqan3/alphabet/concept.hpp (hunk 2 above)
Comment (h-2) [inline include/seqan3/alphabet/concept.hpp]: There are classes that fulfil only this concept, like cartesian_composition

File: include/seqan3/alphabet/concept.hpp (hunk 3 above)
Comment (h-2) [inline include/seqan3/alphabet/concept.hpp]: Hm, I don't know, what do other think? I think the function body is very small, `a` represents allphabet, `i` represents integer and `c` represents char...

File: include/seqan3/alphabet/concept_detail.hpp (hunk 5 above)
Comment (h-2) [inline include/seqan3/alphabet/concept_detail.hpp]: 👍

File: include/seqan3/alphabet/concept_detail.hpp (hunk 7 above)
Comment (h-2) [inline include/seqan3/alphabet/concept_detail.hpp]: It says that it is an implementation of a concept requirement with a link that concept requirement (which has detailed) explanation. OTOH It needs to say *what kind of implementation* / for which types it is → it is the implementation that delegates to a member function. There are and will be other implementations that don't delegate so the distinction is important.

File: include/seqan3/alphabet/concept_detail.hpp (hunk 8 above)
Comment (h-2) [inline include/seqan3/alphabet/concept_detail.hpp]: Hm, I don't understand, the rank of the letter is the rank in every alphabet, or not ❓

File: include/seqan3/alphabet/nucleotide/concept.hpp (hunk 9 above)
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/concept.hpp]: yes, doxygen draws a graph that shows which classes "implement" a concept now.

**Comment by h-2** [comment]
On a more general note: The main documentation page for the alphabet concept is still TODO. It will explain in detail which concepts they are, why we have rank and char representation et cetera so things will be much clearer in the future.
I have it written in my head, but want to focus on more important changes today.

@rrahn can you also review?

File: include/seqan3/alphabet/concept.hpp [hunk 10]
```diff
@@ -43,219 +43,81 @@
...
- * \param c The alphabet letter that you wish to convert to char.
- * \returns A reference to the output stream.
+/*!\interface seqan3::semi_alphabet_concept <>
+ * \brief The basis for seqan3::alphabet_concept, but requires only rank interface (not char).
+ * \ingroup alphabet
+ *
+ * \todo document
+ * This concept is mostly used internally and not relevant fo
+ *
+ * Beyond the requirements stated below, the type needs to a plain old datatype (`std::is_pod_v`)
```
Comment (rrahn) [inline include/seqan3/alphabet/concept.hpp]: missing be?

File: include/seqan3/alphabet/concept.hpp (hunk 3 above)
Comment (rrahn) [inline include/seqan3/alphabet/concept.hpp]: I am also not the fan of to short. Call it **alpha** maybe.

Comment (rrahn) [inline include/seqan3/alphabet/concept.hpp]: and **int** and **ch** 😁

File: include/seqan3/alphabet/concept_detail.hpp [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+// ------------------------------------------------------------------
+
+/*!\typedef template <typename alphabet_type> char_type seqan3::underlying_char_t<>
+ * \brief Implementation of \link char_type seqan3::underlying_char_t<alphabet_concept> \endlink
+ * that delegates to `typename alphabet_type::char_type` [type metafunction].
+ * \tparam alphabet_type Must provide a member type `char_type`.
+ */
+template <typename alphabet_type>
+//!\cond
+    requires requires (alphabet_type a) { typename alphabet_type::rank_type; }
```
Comment (rrahn) [inline include/seqan3/alphabet/concept_detail.hpp]: why do you require the rank_type when you ask for the char_type?

File: include/seqan3/alphabet/concept_detail.hpp (hunk 8 above)
Comment (smehringer) [inline include/seqan3/alphabet/concept_detail.hpp]: Well, although they are of course alphabetically ranked, I would include a documentation like

dna:
- 0 == dna::A
- 1 == dna::C
...

Just for a overview. (especially concerning larger alphabets, or quality and structure alphabets that might not be as obvious)

no ?

Comment (h-2) [inline include/seqan3/alphabet/concept_detail.hpp]: hm, I am not convinced, but in any case, it would be part of polishing the nucleotide submodule and not part of this PR.

Comment (smehringer) [inline include/seqan3/alphabet/concept_detail.hpp]: ok

**Comment by h-2** [comment]
The doxygen is not yet 100% final (because it is a real pain 😢 ), but I have incorporated most of your changes. However some new things were added, too, and some things moved so please review again!

File: include/seqan3/alphabet/detail/member_exposure.hpp [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+
+/*!\brief Implementation of \link rank_type seqan3::to_rank(semi_alphabet_concept const alph) \endlink
+ * that delegates to a member function.
+ * \tparam alphabet_type Must provide a `.to_rank()` member function.
+ * \param alph The alphabet letter that you wish to convert to rank.
+ * \returns The letter's value in the alphabet's rank type (usually a `uint*_t`).
+ */
+template <typename alphabet_type>
+constexpr underlying_rank_t<alphabet_type> to_rank(alphabet_type const alph)
+    requires requires (alphabet_type alph) { { alph.to_rank() } -> underlying_rank_t<alphabet_type>; }
```
Comment (rrahn) [inline include/seqan3/alphabet/detail/member_exposure.hpp]: sometimes you wrap it into a `cond` clause and sometimes not?
